modules/uhttp/router.go

```diff
@@ -22,17 +22,42 @@ package uhttp
 
 import (
 	"net/http"
-
-	"github.com/gorilla/mux"
+	"regexp"
+	"sync"
 
 	"go.uber.org/fx/service"
 )
 
+// The HandlerFunc type is an adapter to allow the use of
+// functions as http handlers.
+type HandlerFunc func(w http.ResponseWriter, r *http.Request)
+
+// ServeHTTP calls f(w, r).
+func (f HandlerFunc) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	f(w, r)
+}
+
+// A RequestMatcher is a matching predicate for an HTTP Request
+type RequestMatcher func(r *http.Request) bool
+
+// PathMatchesRegexp returns a matcher that evaluates the path against the given regex
+func PathMatchesRegexp(re *regexp.Regexp) RequestMatcher {
+	return func(r *http.Request) bool {
+		return re.MatchString(r.URL.Path)
+	}
+}
+
+type route struct {
+	matches RequestMatcher
+	handler http.Handler
+}
+
 // Router is wrapper around gorila mux
 type Router struct {
```

Comment:
I am confused. we are chaining using http.Handler only, except http.Handler doesn't have chaining. Thats why we have middleware to provide the chaining. you can't chain handlers with implementing only http.Handler as is.